README

                                    phpMp
                        http://www.musicpd.org

phpMp is a web interface (written in PHP) for the Music Player Daemon (MPD).

To install phpMp, see INSTALL.

current version phpMp[0.14.3] for PHP/7.3
(orig repo https://github.com/tswaehn/phpmp)
Thanks to [Avuton Olrich][Niccolo Rigacci][Warren Dukes][Sven Ginka]


inline=
----------------------------------------------------------------------------

If you want to use phpMp as a search engine in your browser of choice you
obviously need to do a little URL constuction. This is not an issue as 
phpMp makes it pretty simple, here's how to set it up.

The url will take up to three arguments: inline, search and server

* inline would be the search term, this is required.
* search would be anything you have in $search_fields in config.php or any.
	This is optional and defaults to "any".
* server in the case you have more than one server in config.php, ordered
	by number. This defaults to '0'.

In Konqueror you would set this search URL to:
http://myserver/phpMp/index.php?inline=\{@}&search=any&server=0


stream_browser
--------------------------------------------------------------------------

To use the stream browser updating you need to have 'allow_url_fopen' set 
to 'On' in your php.ini. This is required only if you want to be able to 
download and update the file from within phpMp. Otherwise you could have 
it as a cronjob. The stream browser will not show up (at all) unless the
following file(s) exist.

It needs to go into the cache/ directory with the name stream-icy.xml or 
stream-shout.xml, respectively. While I host a shoutcast xml file, which 
updated tri-daily you can generate your own using contrib/shoutcast.pl, 
which could also be setup as a cronjob.


MiniFAQ
-------

Q. Can you put the Playing Song or the Status in the Topic?

A. Yes and no, Of course, it's possible. It won't happen here because
   it would require the full page to reload rather than just the playlist,
   or javascript.

phpMp is released under the GNU Public License.
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
For the full license, see COPYING.


ChangeLog
ver 0.14.4	-- by tswaehn
	-fix: issue #4
ver 0.14.3	-- by Gerard64
	-fix: issue #1
ver 0.14.2	-- by tswaehn
	-fix: issue #5
ver 0.14.1	-- by pcwalden
	-fix: issue #2 
ver 0.14.0      -- by tswaehn
        - fix: PHP7.0 adjustments
ver 0.12.1	-- by tswaehn
	-fix: changed all "split" to "explode" due to deprecated "split"
	-fix: info2html.php 
			-	line 32: $index was uninitialized
	-bug: info2html.php
		-	line 211: undefined index in $lsinfo due to an entry with index "-1"
		> info.php "getLsInfo();" returns erroneous value for music when playlists come with
		  an additional information 
	-fix: info.php "getLsInfo();"
	-fix: features.php 
		-	line 141: use of date() without setting a timezone
	-fix: features.php 
		-	line 144: "A non well formed numeric value encountered in ..."

	
	
ver 0.12.0 (2006/x/x)

This was near a rewrite, tons of code changed, only the changes that were less significant or that I totally forgot
are not here. Due to changes in this version the way it loads has completely changed.

My experience is, on a 800 mhz machine phpMp 0.11.x loads at 60% for about 1.8 seconds, phpMp 0.12.0 loads at 100%
for about 0.8 seconds, this gets shorter (to a limit) with each load due to caching in linux/php. This is with the
machine with MPD, Apache2 & phpMp serving.

1) Fixed some problems with server change
2) Changed all prints to echos for uniformity
3) Got rid of all *_body.php 
	move playlist_body->playlist
	move main_body->main
	move other *_body to features.php as functions
4) Created server change feature, active server stays at the top
5) Removed all config scalars and created a config array
6) Added anchor titles to most things to help out the n00bs
7) Made playlist frame scalable by centering most text (as config option) and rewriting/adjusting playlist tables.
8) Created crop function, used for playlist
9) Changed delete playlist behaviour
10) Changed update behaviour, now incremental is possible
11) Added config option overwrite_playlists
12) Added (better) comments to config.php
13) Added regex_filename config option to help prevent playlist breakage, and to pretty up the song output
14) Moved printDirectoryTable to info2html.php
15) Completely revamped index.php, now everything is accessed through index.php
16) Removed tons of duplicate php code
17) Remove all /n (newline)s that were used for HTML, use htmltidy from here on out.
18) Added phpMp/MPD version to stats
19) Changed logout behaviour, now a simple logout by just clicking logout
20) AddAll can now add everything in the MPD database
21) Removed seperate find() stuff, now integrated into the feature search()
22) Alternating colors everywhere except stats
23) Removed one ugly hack for crossfade|random|repeat for a CSS hack
24) Frame border between playlist & main is now invisible.
25) Stats is now a table.
26) Much improved password support, If password is needed now it should be detected, you should be sent to password table automatically if login is needed and when logging in user will be sent directly to main menu.
27) Configure options are now much better commented, and it is one array now.
28) Added play_pause config option to alternate between play and pause rather than have two seperate buttons
29) Make config option to align
30) Removed $_GET & $_POST from extract() due to known security problems, now there is a setupGet() that takes two variables, the first variable is an array that hasthe keys from $_GET that you want. The second number is the number of keys in the array from the first variable.
31) Moved connection, connection detection, command execution, includes, login stuff, cookies to the index, so we don't duplicate over frames.
32) Added option to the config to add the current server to the browser display.
29) Make config option to align
30) Don't show the index if sorting by Time because it's not only broken but useless
31) Moved connection, connection detection, command execution, includes, login stuff, cookies to the index, so we don't duplicate over frames.
32) Added option to the config to add the current server to the browser display.
33) Added a icecast/shoutcast browser to the streams feature (thanks goto QballCow for the idea!)
34) Made cookies per-server so when servers are changed the cookies for one server don't conflict with another servers'
35) Status is now shown, rather than just printing 'Playing'
36) Changed use_javascript_add_all to use_javascript
37) Don't let the MPD command 'kill' be used from MPD
38) Removed NoFrames support, if it becomes unbroken _then_ put it in.
39) Remove all underlines
40) Change kbs to kbps
41) Cleaned up printPlaylistInfo
42) Changed pickSort() to a switch type rather than a if..else if
43) Changed hide_threshold behaviour due to a phpMp core bug, now the number you put in will be times two
44) Got rid of most globals, most were unneeded due to structural changes
45) Build command lists then send, that way we max out the block rather than sending partial TCP packets. May see speed difference in playlist due to constant loading.
46) Change from decodeHTML() sanitizeForPost() sanitizeForURL() to rawurlencode() and rawurldecode() integrated php functions
47) Added ability for phpMp to automatically download and add m3u and pls playlists
48) Search now tells how many results were found
49) Genre/Year support were added
50) Oddcast/Streamcast Stream browser added
51) All E_NOTICE warnings were erradicated
52) Added an 'Outputs' menu for enabling/disabling different outputs in MPD
53) When MPD is updating it's visible on the main page now, also 	
54) Prevent updates from colliding
55) Completely implemented commands/notcommands support across the client, this will make things be much more predictable.
56) If a sorting by something, if you click it again the sorting is reversed
57) If filenames_only = "yes" searching now sorts correctly
58) Only show metadata when it exists in the table
59) Redo the way time is calculated in the statistics

index.php
<?php
ob_start();
require "info.php";
require "info2html.php";
require "config.php";
require "utils.php";

/* I have not yet figured out why $_REQUEST is required here, but it is */
$server =	isset( $_REQUEST["server"] )	?	$_REQUEST["server"]	:	"0";
$feature =	isset($_GET["feature"])		?	$_GET["feature"]	:	"";

// This will load the data for a streamurl
if( ! empty( $feature ) && ( strcmp( $feature, "stream-icy" ) == "0" || strcmp( $feature, "stream-shout" ) == "0" ))
{
	require "xml-parse.php";
}

if( sizeof( $servers ) > 1 && strcmp( $config["server_in_title"],"yes" ) == "0" )
{
	$config["title"] .= strlen( $servers[$server][2] ) > "0" ? " ({$servers[$server][2]})" : " ({$servers[$server][0]})";
}

$host = $servers[$server][0];
$port = $servers[$server][1];

// This should keep us from calling status a billion times
//global $status;

// This variable is a argument to $_COOKIE[*] to make it where your cookies
// won't go any old place, but only to the host/port that you are speaking to
$hostport = $host . ":" . $port;

// This will extract the needed GET/POST variables
$arg =		isset( $_GET["arg"])		?	$_GET["arg"]		:	"";
$arg2 =		isset( $_GET["arg2"])		?	$_GET["arg2"]		:	"";
$body =		isset( $_GET["body"])		?	$_GET["body"]		:	"";
$command =	isset( $_GET["command"])	?	$_GET["command"]	:	"";
$feature =	isset( $_GET["feature"] )	?	$_GET["feature"]	:	"";
$remember =	isset( $_REQUEST["remember"] )	?	$_REQUEST["remember"]	:	"";
$passarg =	isset( $_REQUEST["passarg"] )	?	$_REQUEST["passarg"]	:	"";
$stream =	isset( $_GET["stream"] )	?	$_GET["stream"]		:	"";
$streamurl =	isset( $_GET["streamurl"] )	?	$_GET["streamurl"]	:	"";
$inline =	isset( $_GET["inline"] )	?	$_GET["inline"]		:	"";
$streamfilter =	isset( $_GET["streamfilter"] )	?	$_GET["streamfilter"]	:	"";
$sfa =		isset($_GET["sfa"])		?	$_GET["sfa"]		:	"";

if(isset($_POST["body"])) {
	$body = $_POST["body"];
	$add_all = isset( $_POST["add_all"] ) ? $_POST["add_all"] : "" ;
}

// $inline is a simple searcher
if( isset($inline)) {
	$search = isset($_GET["search"]) ? $_GET["search"] : "";
}

if( ! empty( $body ))
{
	if( strcmp( $body, "main" ) == "0" )
	{
		if( ! empty($inline)) {
			$arg = $inline;
			$feature = "search";
			if(empty($search) || ! isset($search)) {
				$search = "any";
			}
			$search_fields = $config["display_fields"];
		}
		else if( ! empty( $feature ))
		{
			if( strcmp( $feature, "search" ) == "0" )
			{
				$search = isset( $_GET["search"] ) ? $_GET["search"] : "";
				$find = isset( $_GET["find"] ) ? $_GET["find"] : "";
				$search_fields = $config["display_fields"];
			}
		}
		$delete = isset( $_GET["delete"] ) ? $_GET["delete"] : "no";
		$dir = isset( $_GET["dir"] ) ? $_GET["dir"] : "";
		$save = isset( $_GET["save"] ) ? $_GET["save"] : "";
		$ordered = isset( $_GET["ordered"] ) ? $_GET["ordered"] : "";
	}
	else if( strcmp( $body, "playlist" ) == "0" )
	{
		// Playlist Hiding Stuff
		if( isset( $_GET["hide"] ))
		{
			$hide = $_GET["hide"];
		}
		else if( strcmp($config["use_cookies"], "yes" ) == "0" && isset( $_COOKIE["phpMp_playlist_hide"][$hostport] ))
		{
			$hide = $_COOKIE["phpMp_playlist_hide"][$hostport];
		}

		if( isset( $hide ) && strcmp( $config["use_cookies"], "yes" ) == "0" )
		{
			setcookie("phpMp_playlist_hide[$hostport]", $hide);
		}
		else
		{
			$hide = 1;
		}

		$show_options = isset( $_GET["show_options"] ) ? $_GET["show_options"] : "0";
	}
}
else
{
	$logout = isset( $_GET["logout"] ) ? $_GET["logout"] : "";
}

header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
header("Cache-Control: no-store, no-cache, must-revalidate");
header("Pragma: no-cache");
header("Content-Type: text/html; charset=UTF-8");

echo "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\">";
echo "<html><head><link rel=\"shortcut icon\" href=\"mpd-favicon.ico\">";

// Open the connection
$fp = fsockopen( $host, $port, $errno, $errstr, 10 );

// If there's no connection, and servers exist goto the server menu
if(! is_resource( $fp ))
{
	if(isset($servers) && (sizeof($servers) > 1))
	{
		include "features.php";
		server( $servers, $host, $port, $colors["server"] );
	}
	die("{$errstr} ({$errno})<br>");
}

// Lets go ahead and get the MPD version while we can
$MPDversion = initialConnect($fp);
$MPDversion = trim($MPDversion);

// Password stuff
if( !empty( $logout ))
{
	setcookie( "phpMp_password[$hostport]", "" );
}
else if( isset( $_COOKIE["phpMp_password"][$hostport] ))
{
	$passarg = $_COOKIE["phpMp_password"][$hostport];
}

if( strlen( $passarg ) > "0" )
{
	$has_password = 1;
	fputs( $fp, "password \"$passarg\"\n" );
	while ( ! feof( $fp ))
	{
		$got = fgets( $fp, 1024 );
		if( strncmp( "OK", $got, strlen( "OK" )) == "0" )
		{
			if( isset( $remember ) && strcmp( $remember, "true" ) == "0" )
			{
				setcookie( "phpMp_password[$hostport]", $passarg, time()+60*60*24*30 );
			}
			else if( ! $_COOKIE["phpMp_password"][$hostport] )
			{
				setcookie( "phpMp_password[$hostport]", $passarg );
			}
			break;
		}
		if( strncmp( "ACK", $got, strlen( "ACK" )) == "0" )
		{
			echo "Password Incorrect, press back button";
			break;
		}
	}
}

if( ! isset( $has_password ))
{
	$has_password = 0;
}

$commands = getCommandInfo( $fp, $MPDversion );
if( $commands["status"] == "1" )
{
	$status = getStatusInfo( $fp );
}

if( ! empty( $command ))
{
	doCommand( $fp, $arg, $arg2, $command, $config["overwrite_playlists"], $status );
	$status = getStatusInfo( $fp ); 
}

// This needs to go down here to give the cookies, server time to load
include "theme.php";
if( $commands["status"] == "1" && $config["smart_updating"] == "yes" && ! empty($status["time"])) {
	$var1 = strtok($status["time"],":");
	$var2 = strrchr($status["time"],":");
	$var2 = str_replace(":","",$var2);
	$sub = ($var2-$var1);
	if($sub<$config["refresh_freq"] && $sub>0) {
		$config["refresh_freq"] = ($var2-$var1);
	}
}

// There might be more that's would prevent phpMp from loading, rather than taking time to figure it out, we'll wait for reports.
if( $commands["listall"] == "0" || $commands["lsinfo"] == "0" || $commands["playlist"] == "0" || $commands["playlistinfo"] == "0" || $commands["stats"] == "0" )
{
	include "features.php";
	setcookie( "phpMp_password[$hostport]", "" );
	unset( $has_password );

	echo "<b>Error:</b> Can't load phpMp due to not having permission to the following commands: ";
	if( $commands["listall"] == "0" )
	{
		echo "listall ";
	}
	if( $commands["lsinfo"] == "0" )
	{
		echo "lsinfo ";
	}
	if( $commands["playlist"] == "0" )
	{
		echo "playlist ";
	}
	if( $commands["playlistinfo"] == "0" )
	{
		echo "playlistinfo";
	}

	echo "<br>";

	login( $fp, $config, $colors["login"], $server, $arg, $dir, $remember );
	server( $servers, $host, $port, $colors["server"], $config, $commands );
}
// This will serve as our front page if called w/o $body
else if( empty( $body ) && empty( $feature ))
{
	unset( $hostport );
	echo "<title>{$config["title"]}</title>";
	echo "</head>";

	echo "<frameset {$config["frames_layout"]}>";
	$mainurl = "index.php?body=main&amp;server=$server";
	if($inline) {
		$mainurl .= "&amp;inline=$inline&amp;search=$search";
	}

	echo "<frame name=\"main\" src=\"$mainurl\" frameborder={$config["frame_border_size"]}>";
	echo "<frame name=\"playlist\" src=\"index.php?body=playlist&amp;server=$server\" frameborder=0>";
	echo "<noframes>NO FRAMES ... try phpMp+</noframes>";
	echo "</frameset>";
}
else
{
	unset( $hostport );
	if( strcmp( $body, "playlist" ) == "0" )
	{
		echo "<META HTTP-EQUIV=\"REFRESH\" CONTENT=\"{$config["refresh_freq"]};URL=index.php?body=playlist&amp;hide=$hide&amp;show_options=$show_options&amp;server=$server\">";
	}
	if( isset( $status["updating_db"] ) && strcmp( $body, "main" ) == "0" )
	{
		$sort =	isset( $_GET["sort"] ) ? $_GET["sort"] : $config["default_sort"];
		echo "<META HTTP-EQUIV=\"REFRESH\" CONTENT=\"{$config["refresh_freq"]};URL=index.php?body=main&amp;sort=$sort&amp;dir=$dir&amp;ordered=$ordered&amp;server=$server\">";
	}
	echo "<title>{$config["title"]} - $body</title>";

	// I would _much_ rather have a php generated stylesheet
	echo "<style type=\"text/css\">";
	echo "* { font-family: {$fonts["all"]}; }";
	echo "A:link, A:visited, A:active { text-decoration: none; border-style: none none none none; }";
	echo "a.green:link, a.green:active, a.green:visited, a.green:hover {background: {$colors["playing"]["on"]}}";
	echo "table { width: 100%; border-style: none }";
	echo "form { padding: 0; margin: 0 }";
	echo "tr.noborder td { border: none; border-style: none; border-color: {$colors["directories"]["body"][0]} }";
	echo "</style>";
	echo "</head>";

	echo "<body link={$colors["links"]["link"]} ";
	echo "vlink={$colors["links"]["visual"]} "; 
	echo "alink={$colors["links"]["active"]} ";
	echo "bgcolor={$colors["background"]}>";

	echo "<!-- The Header (index.php) Ends Here, Body begins here -->";

	include $body . ".php";

	echo "</body>";
}
echo "</html>";
fclose( $fp );
ob_end_flush();
?>
